### PR TITLE
add EvaluationBackgroundTask

### PR DESCRIPTION
**Pull request by yshrsmz**
- added background evaluation polling
- added sample project

### CODE REVIEW

File: bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EvaluationBackgroundTask.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import io.bucketeer.sdk.android.internal.loge
+import io.bucketeer.sdk.android.internal.logi
+
+internal class EvaluationBackgroundTask : BroadcastReceiver() {
+  override fun onReceive(context: Context?, intent: Intent?) {
+    // As we can't serialize logger we need to use BKTClient singleton.
+    // This means
+    // - BKTClient#initialize must be called first
+    // - background polling happens as long as BKTClient singleton lives
+    val client = try {
```
Comment (yshrsmz) [inline bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EvaluationBackgroundTask.kt:22]: At first, I thought we need to save config and user to run background task, but it turned out that we can't save logger.

So currently background task implementation relies on BKTClient singleton.

Do you have any idea? I think this implementation covers basic use cases.

File: sample2/src/main/java/io/bucketeer/sample/App.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import io.bucketeer.sdk.android.BKTConfig
+import io.bucketeer.sdk.android.BKTUser
+import io.bucketeer.sdk.android.BuildConfig
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import java.util.concurrent.TimeUnit
+
+class App : Application(), LifecycleObserver {
```
Comment (yshrsmz) [inline sample2/src/main/java/io/bucketeer/sample/App.kt:18]: Added a simple sample app by copying & pasting `sample` module.

**Comment by yshrsmz** [comment]
@cre8ivejp do we need to send EventBackgroundTask as well? I think we don't because evaluations requests in the background make metrics events as usual, and the events queue eventually exceeds max count(and send events).

**Comment by cre8ivejp** [comment]
> do we need to send EventBackgroundTask as well?

@yshrsmz, we need to keep sending events to the server. So we need to check every 30 seconds if there are new events or if it exceeds the max size.
It will be fine if the event foreground task keeps working in the background.

**Comment by yshrsmz** [comment]
@cre8ivejp 

> we need to check every 30 seconds if there are new events or if it exceeds the max size.

Ok, I'm going to add the background task for events as well.

File: bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EvaluationBackgroundTask.kt (hunk 1 above)
Comment (yshrsmz) [reply bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EvaluationBackgroundTask.kt:22]: If a user interacts with `BKTClient#stringVariation()` and others, then BKTClient should be alive.

**Comment by yshrsmz** [comment]
@cre8ivejp 

Added EventBackgroundTask.

There's one concern:

- AlarmManager, which we use as a way to run background tasks, does not accept an interval shorter than 1 minute.
  - This means we can't send events every 30 seconds.
  - Do we use force set 1 minute for background event sync? or add `BKTConfig#backgroundEventsFlushInterval` with min value check?

**Comment by yshrsmz** [comment]
Speaking about background tasks in general, there's no easy way to run a job every minute or in a shorter interval.

WorkManager's minimum interval is 15 minutes.
AlarmManager is 1 minute(but not the exact time)
ForegroundService can handle this use case, but it needs notification to run.

if [this](https://github.com/ca-dp/bucketeer-android-sdk/pull/72#discussion_r970602390) is OK, we can choose to use Executor in background as well, but this is not that performant.

**Comment by cre8ivejp** [comment]
What do you think about doing this?
We use AlertManager for both cases.

### EventTask

- We use 1 minute for both foreground and background tasks
- Before starting the background task, we flush the events

### EvaluationTask

- For foreground tasks, we use 10 minutes by default, but check for a minimum interval, in this case, 1 minute. Anything sooner than that, we set it to 1 minute (Like you implemented it before)
- For background tasks, we use 1 hour by default, but check for a minimum interval, in this case, 20 minutes. Anything sooner than that, we set it to 20 minutes.
For background tasks, we would also have an option to use the WorkManager if needed.

**Comment by yshrsmz** [comment]
Thanks, that sounds good.

I will add checks in BKTConfig and other related classes.

> For background tasks, we would also have an option to use the WorkManager if needed.

Yes, but then we need to add one more dependency, and I'd like to align implementations if possible(In this case, I prefer to use AlarmManager for both background tasks).

**Comment by cre8ivejp** [comment]
Oh, I didn't know the WorkManager needs another dependency.
Let's go with AlertManager!
Thank you.

File: bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/BKTConfig.kt [hunk 3]
```diff
@@ -78,6 +82,21 @@ data class BKTConfig internal constructor(
       require(this.endpoint?.toHttpUrlOrNull() != null) { "endpoint is invalid" }
       require(!this.featureTag.isNullOrEmpty()) { "featureTag is required" }
 
+      if (this.pollingInterval < MINIMUM_POLLING_INTERVAL_MILLIS) {
```
Comment (yshrsmz) [inline bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/BKTConfig.kt:85]: added validations

File: bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/TaskScheduler.kt [hunk 4]
```diff
@@ -6,26 +6,51 @@ import io.bucketeer.sdk.android.internal.di.Component
...
 
   // to background
   override fun onStop(owner: LifecycleOwner) {
+    // stop foreground tasks
     foregroundSchedulers.forEach { it.stop() }
 
-    // TODO: start background tasks
+    // flush events before switching to background tasks
+    executor.execute {
+      component.eventInteractor.sendEvents(force = true)
```
Comment (yshrsmz) [inline bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/TaskScheduler.kt:45]: flush events while switching to background tasks

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -57,6 +58,8 @@ materialComponents = { module = "com.google.android.material:material", version
 
 stetho = { module = "com.facebook.stetho:stetho", version = "1.6.0" }
 
+kotlin-coroutines-android = { module = "org.jetbrains.kotlinx:kotlinx-coroutines-android", version.ref = "coroutines" }
```
Comment (yshrsmz) [inline gradle/libs.versions.toml:61]: kotlin coroutines is used only in sample

File: bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EvaluationBackgroundTask.kt [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import io.bucketeer.sdk.android.internal.util.createBroadcastPendingIntent
+import io.bucketeer.sdk.android.internal.util.getAlarmManager
+
+internal class EvaluationBackgroundTask : BroadcastReceiver() {
+  override fun onReceive(context: Context?, intent: Intent?) {
+    // We need to use BKTClient singleton because we can't serialize logger
+    // This means
+    // - BKTClient#initialize must be called first
+    // - background polling happens as long as BKTClient singleton lives
+    val client = try {
```
Comment (yshrsmz) [inline bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EvaluationBackgroundTask.kt:22]: https://github.com/ca-dp/bucketeer-android-sdk/pull/72#discussion_r970602390

File: bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EventBackgroundTask.kt [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    } catch (e: Throwable) {
+      null
+    }
+
+    if (client == null) {
+      logd { "BKTClient is not initialized, skipping background Event sync..." }
+      return
+    }
+
+    val pendingResult = goAsync()
```
Comment (yshrsmz) [inline bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EventBackgroundTask.kt:33]: Async BroadcastReceiver can run for up to 10 seconds, but that should be enough in most cases.

https://developer.android.com/reference/android/content/BroadcastReceiver

**Review by yshrsmz — COMMENTED** [review]
@cre8ivejp should be ready now.
Please take a look 🙏

File: bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EventBackgroundTask.kt [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    }
+  }
+
+  class Scheduler(
+    private val context: Context,
+    private val interval: Long,
+  ) : ScheduledTask {
+    override fun start() {
+      stop()
+      logi { "start background vnt polling...: $interval" }
```
Comment (cre8ivejp) [inline bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EventBackgroundTask.kt]: nit: `vnt` -> `Event`

File: sample2/src/main/java/io/bucketeer/sample/Constants.kt [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package io.bucketeer.sample
+
+/**
+ * Created by Alessandro Yuichi Okimoto on 2019-12-19.
```
Comment (cre8ivejp) [inline sample2/src/main/java/io/bucketeer/sample/Constants.kt]: Please delete this comment from the header.

File: bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EventBackgroundTask.kt (hunk 8 above)
Comment (yshrsmz) [reply bucketeer2/src/main/kotlin/io/bucketeer/sdk/android/internal/scheduler/EventBackgroundTask.kt]: thanks, fixed [`27cd911` (#72)](https://github.com/ca-dp/bucketeer-android-sdk/pull/72/commits/27cd91145b79a46936f7e7adc0d4a92b42e0774e)

File: sample2/src/main/java/io/bucketeer/sample/Constants.kt (hunk 9 above)
Comment (yshrsmz) [reply sample2/src/main/java/io/bucketeer/sample/Constants.kt]: thanks, removed [`74e8f81` (#72)](https://github.com/ca-dp/bucketeer-android-sdk/pull/72/commits/74e8f814b9da40222c388ab458977597d63b94da)

**Review by cre8ivejp — APPROVED** [review]
Amazing work!
Thank you!

**Comment by yshrsmz** [comment]
thanks!

**Pull request merged**